### ISSUE TITLE
Enable adapters to provide a URL instead of a read stream

### ISSUE THREAD

**Issue by loanlaux** [reactioncommerce/reaction-file-collections#14]
Given that storage adapters are expected to provide a read stream on file retrievals, the current state of the S3 adapter is that it fetches files from the S3 SDK and pushes them to a read stream that is made available by `_getReadStream`.

However, when using S3 to store images, users will generally want to refer to their images by either the S3 URL or a CDN pointing to S3. Having their Reaction instances fetch the files and make them available to the user usually defeats the purpose of using an external storage solution.

It would be great if storage adapters had the option to provide ready-to-use URLs instead of read streams. The logical translation of this could look like "if `_getReadStream` is `undefined`, then use `_getReadUrl`".

**Comment by loanlaux** [comment]
Alternative option: when calling `someFile.url()`, reaction-file-collections could check if the media object in Mongo has a `url` field set. If yes, then this `.url()` call returns the stored URL and doesn't use the `_getReadStream` method.

**Comment by aldeed** [comment]
@loanlaux Yes, I think it could take place in the `url` function, and each store object could optionally have an `externalUrl` prop set by the adapter.

You will have to consider security, though. If there are read permission checks in the file-collections implementation, they would be bypassed by putting `externalUrl` directly on the object. Also, the client would then have to handle providing appropriate authentication to the S3 server / cloudfront. So in general, maybe this should be behind some kind of `isPublic` S3 adapter constructor option, where the S3 adapter will only set `externalUrl` if `isPublic` is true.

**Comment by aldeed** [comment]
I am thinking it would be the full S3 URL. So near the top of the existing `FileRecord#url`, you would add

```js
const externalUrl = fileRecord.externalUrl({ store });
if (externalUrl) return externalUrl;
```

Where `externalUrl` function should return that prop from the store object, similar to how `name` and such work. And the storage adapter could optionally set that prop to the full URL when it emits the "stored" event.

By default the adapter `_getWriteStream` could do public bucket URL + key to make the full `externalUrl`, but you could also make the base URL overridable with a constructor option, allowing someone to pass in their cloudfront base URL.

**Comment by loanlaux** [comment]
Got it, that makes sense! I'll PR this if that's okay. Or do you want to take care of it?

**Comment by aldeed** [comment]
Feel free. I won't have time for awhile.

**Comment by jhonsfran1165** [comment]
Hi @loanlaux I'm not sure if I understood well but this PR works for me. Let me know what you think. 

https://github.com/reactioncommerce/reaction-file-collections/pull/47

**Comment by CristianCucunuba** [comment]
It would helpful now to support this, to use it in the new storefront version

**Comment by focusaurus** [comment]
I'm going to close this issue as @jhonsfran1165 's commits landed in #49. Please reopen if there's anything further we need to address.
